File: install/extract_text_from_image.swift
#!/usr/bin/env swift

import Foundation
import Vision

func writeError(_ message: String) {
    let data = (message + "\n").data(using: .utf8)!
    FileHandle.standardError.write(data)
}

guard CommandLine.arguments.count == 2 else {
    writeError("Usage: extract_text_from_image.swift <image_path>")
    exit(1)
}

let imagePath = CommandLine.arguments[1]
guard FileManager.default.fileExists(atPath: imagePath) else {
    writeError("Error: Image file does not exist: \(imagePath)")
    exit(1)
}

guard let imageURL = URL(fileURLWithPath: imagePath) as URL?,
      let imageData = try? Data(contentsOf: imageURL) else {
    writeError("Error: Failed to load image")
    exit(1)
}

let requestHandler = VNImageRequestHandler(data: imageData, options: [:])
let textRequest = VNRecognizeTextRequest()
textRequest.recognitionLevel = .fast

// Configure recognition languages to support all available languages
// Supported languages as of macOS 13.0/iOS 16.0+:
// English, French, Italian, German, Spanish, Portuguese (Brazil),
// Chinese (Simplified/Traditional), Cantonese (Simplified/Traditional),
// Korean, Japanese, Russian, Ukrainian
// Note: To get the exact list at runtime, use:
// try textRequest.supportedRecognitionLanguages(for: .fast, revision: VNRecognizeTextRequestRevision1)
textRequest.recognitionLanguages = [
    "en-US",    // English
    "fr-FR",    // French
    "it-IT",    // Italian
    "de-DE",    // German
    "es-ES",    // Spanish
    "pt-BR",    // Portuguese
    "zh-Hans",  // Simplified Chinese
    "zh-Hant",  // Traditional Chinese
    "yue-Hans", // Simplified Cantonese
    "yue-Hant", // Traditional Cantonese
    "ko-KR",    // Korean
    "ja-JP",    // Japanese
    "ru-RU",    // Russian
    "uk-UA",    // Ukrainian
]

// Enable automatic language detection as fallback
// This helps when the text contains multiple languages or languages not in the list above
textRequest.automaticallyDetectsLanguage = true

do {
    try requestHandler.perform([textRequest])
} catch {
    writeError("Error: Vision framework request failed: \(error)")
    exit(1)
}

guard let observations = textRequest.results, !observations.isEmpty else {
    // No text found - exit with code 1 but no error message (this is expected)
    exit(1)
}

var extractedTextParts: [String] = []
for observation in observations {
    let topCandidates = observation.topCandidates(1)
    guard let topCandidate = topCandidates.first else {
        continue
    }
    extractedTextParts.append(topCandidate.string)
}

let extractedText = extractedTextParts.joined(separator: " ")
if extractedText.isEmpty {
    exit(1)
}

print(extractedText)

File: install/extract_text_from_image.py
#!/usr/bin/env python3
"""
Extract text from an image using EasyOCR.
Similar to install/extract_text_from_image.swift for macOS, but uses EasyOCR instead of Vision framework.
"""

import sys
import os

def write_error(message: str) -> None:
    """Write error message to stderr."""
    sys.stderr.write(f"{message}\n")
    sys.stderr.flush()

def main() -> int:
    """Main function to extract text from image."""
    # Check command-line arguments
    if len(sys.argv) != 2:
        write_error("Usage: extract_text_from_image.py <image_path>")
        return 1
    
    image_path = sys.argv[1]
    
    # Verify image file exists
    if not os.path.exists(image_path):
        write_error(f"Error: Image file does not exist: {image_path}")
        return 1
    
    try:
        # Import EasyOCR
        import easyocr
    except ImportError:
        write_error("Error: easyocr module not found. Please install it with: pip install easyocr")
        return 1
    
    try:
        # Initialize EasyOCR reader
        # This will download models on first use (may be slow)
        # EasyOCR supports 80+ languages. Common language codes:
        # en (English), fr (French), de (German), es (Spanish), it (Italian),
        # pt (Portuguese), ch_sim (Simplified Chinese), ch_tra (Traditional Chinese),
        # ja (Japanese), ko (Korean), ru (Russian), uk (Ukrainian), th (Thai), etc.
        # See https://www.jaided.ai/easyocr/ for full list
        # gpu=False uses CPU only (faster install, no CUDA dependencies)
        languages = [
            'en',      # English
            'ch_tra',  # Traditional Chinese
        ]
        reader = easyocr.Reader(languages, gpu=False)
        
        # Read text from image
        results = reader.readtext(image_path)
        
        # Extract text from results
        # results is a list of tuples: (bbox, text, confidence)
        extracted_text_parts = []
        for (bbox, text, confidence) in results:
            extracted_text_parts.append(text)
        
        # Join all text parts with spaces
        extracted_text = " ".join(extracted_text_parts)
        
        if not extracted_text.strip():
            # No text found - exit with code 1 but no error message (this is expected)
            return 1
        
        # Output extracted text to stdout
        print(extracted_text)
        return 0
        
    except Exception as e:
        write_error(f"Error: EasyOCR text extraction failed: {e}")
        return 1

if __name__ == "__main__":
    sys.exit(main())
